pkg/controller/factory/base_controller.go

```diff
@@ -62,6 +65,39 @@ func (s *scheduledJob) Run() {
 	s.queue.Add(DefaultQueueKey)
 }
 
+func waitForPreconditions(ctx context.Context, controllerName string, preconditions ...PreconditionFn) error {
+	if len(preconditions) == 0 {
+		return nil
+	}
+
+	var (
+		wg       sync.WaitGroup
+		errs     []error
+		errsLock sync.Mutex
+	)
+
+	wg.Add(len(preconditions))
+	for _, preconditionFunc := range preconditions {
+		go func(preconditionCtx context.Context, fn PreconditionFn) {
```

Comment:
ctx

Reply:
no, it can collide with upper level `ctx`